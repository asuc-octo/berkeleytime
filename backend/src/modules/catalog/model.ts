import mongoose, { Schema, InferSchemaType, Types } from "mongoose";

export const Identifier = {
  type: {
    type: String
  }, 
  id: String
}

export const CourseSchema = new Schema({
  _id: {
    type: Types.ObjectId,
    required: true
  },
  identifiers: [Identifier],
  _created: {
    $date: {
      $numberLong: String
    }
  },
  _updated: {
    $date: {
      $numberLong: String
    }
  },
  _version: Number,
  academicCareer: {
    code: String,
    description: String
  },
  academicGroup: {
    code: String,
    description: String
  },
  academicOrganization: {
    code: String,
    description: String
  },
  allowMultipleEnrollments: Boolean,
  anyFeesExist: Boolean,
  blindGrading: Boolean,
  catalogNumber: {
    prefix: String,
    number: Date,
    suffix: String,
    formatted: String
  },
  cip: {
    code: String,
    description: String
  },
  classDisplayName: {
    type: String,
    required: true
  },
  classSubjectArea: {
    code: String,
    description: String
  },
  contactHours: Number,
  createdDate: Date,
  credit: {
    type: String,
    value: {
      fixed: {
        units: Number
      }
    }
  },
  creditRestriction: {
    restrictionText: String,
    restrictionCourses: {
      creditRestrictionCourses: [{
        course: {
          identifiers: [Identifier],
          displayName: String
        }
      }]
    }
  },
  crossListing: {
    count: Number,
    courses: [String]
  },
  departmentNicknames: String,
  description: String,
  displayName: String,
  finalExam: {
    code: String,
    description: String
  },
  formatsOffered: {
    description: String,
    formats: [{
      termsAllowed: {
        termNames: [String]
      },
      sessionType: Date,
      description: String,
      aggregateMinContactHours: Number,
      aggregateMaxContactHours: Number,
      minWorkloadHours: Number,
      maxWorkloadHours: Number,
      anyFeesExist: Boolean,
      components: [{
        instructionMethod: {
          code: String,
          description: String
        },
        primary: Boolean,
        minContactHours: Number,
        maxContactHours: Number,
        feesExist: Boolean,
        finalExam: {
          code: String,
          description: String
        }
      }]
    }],
    typicallyOffered: {
      terms: {
        termNames: [String]
      },
      comments: String
    },
    summerOnly: Boolean
  },
  formerDisplayName: String,
  fromDate: Date,
  gradingBasis: {
    code: String,
    description: String,
  },
  hegis: {
    code: String,
    description: String,
  },
  instructorDropConsentRequired: Boolean,
  multipleTermNumber: Number,
  preparation: {
    requiredText: String,
    requiredCourses: {
        courses: {
            identifiers: [Identifier],
            displayName: String,
        }
    }
  },
  primaryInstructionMethod: {
    code: String,
    description: String
  },
  printInCatalog: Boolean,
  printInstructors: Boolean,
  proposedInstructors: [String],
  repeatability: {
    repeatable: Boolean
  },
  spansMultipleTerms: Boolean,
  status: {
    code: String,
    description: String
  },
  subjectArea: {
    code: String,
    description: String
  },
  tie: {
    code: String,
    description: String,
  },
  title: {
    type: String,
    required: true
  },
  toDate: Date,
  transcriptTitle: String,
  updatedDate: Date,
  workloadHours: Number,
})

export const ClassSchema = new Schema({
  _id: {
    type: Types.ObjectId,
    required: true
  }
  
})

export const CourseModel = mongoose.model("sis_course", CourseSchema, "sis_course");
export type CourseType = InferSchemaType<typeof CourseSchema>;