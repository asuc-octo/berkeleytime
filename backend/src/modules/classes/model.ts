import mongoose, { Schema, InferSchemaType, Types } from "mongoose";

export const SisClassSchema = new Schema({
  _id: {
    type: Types.ObjectId,
  },
  course: {
    catalogNumber: {
      number: {
        type: String,
      },
      formatted: {
        type: String,
      },
    },
    displayName: {
      type: String,
    },
    identifiers: {
      type: [],
    },
    subjectArea: {
      code: {
        type: String,
      },
      description: {
        type: String,
      },
    },
    title: {
      type: String,
    },
    transcriptTitle: {
      type: String,
    },
  },
  number: {
    type: Date,
  },
  session: {
    id: {
      type: String,
    },
    name: {
      type: String,
    },
    term: {
      id: {
        type: Date,
      },
      name: {
        type: String,
      },
    },
  },
  _created: {
    type: Date,
  },
  _updated: {
    type: Date,
  },
  _version: {
    type: Number,
  },
  aggregateEnrollmentStatus: {
    status: {
      code: {
        type: String,
      },
      description: {
        type: String,
      },
    },
    enrolledCount: {
      type: Number,
    },
    minEnroll: {
      type: Number,
    },
    maxEnroll: {
      type: Number,
    },
    maxWaitlist: {
      type: Number,
    },
    instructorAddConsentRequired: {
      type: Boolean,
    },
    instructorDropConsentRequired: {
      type: Boolean,
    },
    waitlistedCount: {
      type: Number,
    },
  },
  allowedUnits: {
    minimum: {
      type: Number,
    },
    maximum: {
      type: Number,
    },
    forAcademicProgress: {
      type: Number,
    },
    forFinancialAid: {
      type: Number,
    },
  },
  anyFeesExist: {
    type: Boolean,
  },
  anyPrintInScheduleOfClasses: {
    type: Boolean,
  },
  anyPrintInstructors: {
    type: Boolean,
  },
  assignedClassMaterials: {
    status: {},
    noneAssigned: {
      type: Boolean,
    },
  },
  blindGrading: {
    type: Boolean,
  },
  contactHours: {
    type: Number,
  },
  displayName: {
    type: String,
  },
  finalExam: {
    code: {
      type: String,
    },
    description: {
      type: String,
    },
  },
  gradingBasis: {
    code: {
      type: String,
    },
    description: {
      type: String,
    },
  },
  instructionMode: {
    code: {
      type: String,
    },
    description: {
      type: String,
    },
  },
  offeringNumber: {
    type: Number,
  },
  primaryComponent: {
    code: {
      type: String,
    },
    description: {
      type: String,
    },
  },
  status: {
    code: {
      type: String,
    },
    description: {
      type: String,
    },
  },
});

export const SisClassHistorySchema = new Schema({
  _id: { type: Types.ObjectId, required: true },
  _createdAt: String,
  _updatedAt: String,
  _version: Number,
  collectionName: String,
  collectionId: { type: Types.ObjectId, required: true },
  diff: {},
});
