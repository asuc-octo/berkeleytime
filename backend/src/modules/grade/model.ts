import mongoose, { Schema, InferSchemaType, Types } from "mongoose";

export const GradeSchema = new Schema({
    _id: { type: Types.ObjectId, required: true },
    _created: String,
    _updated: String,
    _version: Number,
    term: {
        year: Number,
        month: Number,
        semester: String,
    },
    CourseControlNbr: Number,
    CourseNbr: String,
    CourseSubjectShortNm: String,
    CourseTitleNm: String,
    EnrollmentCnt: Number,
    GradeNm: String,
    GradeSortNbr: String,
    GradeSubtypeDesc: String,
    GradeTypeDesc: String,
    InstructorName: [String],
    SectionNbr: Number
});

export const GradeModel = mongoose.model("calanswers_grade", GradeSchema, "calanswers_grade");
export type GradeType = InferSchemaType<typeof GradeSchema>;
