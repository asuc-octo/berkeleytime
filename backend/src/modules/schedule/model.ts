import { ObjectId } from "mongodb";
import mongoose, { Schema, InferSchemaType, Types } from "mongoose";
import { User } from "../user/fixture";
import { CustomEvent } from "./fixture";

export const ScheduleSchema = new Schema({
  name: { type: String, required: false },
  created_by: { type: ObjectId, required: true },
  date_created: { type: String, required: true },
  last_updated: { type: String, required: true },
  term: { type: String, required: true },
  public: { type: Boolean, required: true },
  class_IDs: { type: [String], required: false },
  section_Ids: { type: [String], required: false },
  custom_events: { type: [CustomEvent], required: false }
});

export const ScheduleModel = mongoose.model("schedule", ScheduleSchema, "schedule");
export type ScheduleType = InferSchemaType<typeof ScheduleSchema>;
