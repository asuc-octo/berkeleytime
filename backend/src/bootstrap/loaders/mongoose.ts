import mongoose from "mongoose";

import { config } from "../../config";

// Close the Mongoose default connection is the event of application termination
process.on("SIGINT", async () => {
  await mongoose.connection.close();
  process.exit(0);
});

export const mongoDBConfig = {
  useNewUrlParser: true,
  useUnifiedTopology: true,
};

// Your Mongoose setup goes here
export default async (): Promise<mongoose.Mongoose> =>
  mongoose.connect(config.mongoDB.uri, mongoDBConfig);
