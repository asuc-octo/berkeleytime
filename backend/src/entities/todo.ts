import { ObjectType, Field, ID } from "type-graphql";
import { prop } from "@typegoose/typegoose";
import { TimeStamps } from "@typegoose/typegoose/lib/defaultClasses";
import { ObjectId } from "mongodb";

@ObjectType()
export class Todo extends TimeStamps {
  @Field()
  readonly _id: ObjectId;

  @Field(() => Date)
  createdAt: Date;

  @Field(() => Date)
  updatedAt: Date;

  @prop()
  @Field()
  content!: string;

  @prop({ default: false })
  @Field()
  isDone!: boolean;
}
