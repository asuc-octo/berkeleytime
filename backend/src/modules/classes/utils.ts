import { EnrollmentInfo } from "../../generated-types/graphql";
import { SisClassHistory } from "./model";

/**
 * Construct enrollment info from SisClassHistory. Populates missing fields of the current
 * SisClass with the last known value.
 *
 * @param sisClassHistory Mongoose history model
 * @returns EnrollmentInfo gql schema
 */
export function constructEnrollment(
  sisClassHistory: SisClassHistory[]
): EnrollmentInfo[] {
  const enrollmentInfo: EnrollmentInfo[] = [];

  let enrolledMax = 0;
  let waitlistedMax = 0;
  let enrolledCount = 0;
  let waitlistedCount = 0;

  for (const classHistory of sisClassHistory) {
    enrolledMax =
      classHistory.diff?.aggregateEnrollmentStatus?.maxEnroll[1] || enrolledMax;
    waitlistedMax =
      classHistory.diff?.aggregateEnrollmentStatus?.maxWaitlist[1] ||
      waitlistedMax;
    enrolledCount =
      classHistory.diff?.aggregateEnrollmentStatus?.enrolledCount[1] ||
      enrolledCount;
    waitlistedCount =
      classHistory.diff?.aggregateEnrollmentStatus?.waitlistedCount[1] ||
      waitlistedCount;

    enrollmentInfo.push({
      enrolledMax,
      waitlistedMax,
      enrolledCount,
      waitlistedCount,
      date: classHistory.createdAt,
    });
  }

  return enrollmentInfo;
}
