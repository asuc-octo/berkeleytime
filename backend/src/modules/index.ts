import { Resolvers } from "../generated-types/graphql";
import User from "./user";
import Schedule from "./schedule";

const modules = [User, Schedule];

// Important: Add all your module's resolver in this
export const resolvers: Resolvers = modules.reduce(
  (acc, module) => ({ ...acc, ...module.resolver }),
  {}
);

export const typeDefs = modules.map((module) => module.typeDef);
