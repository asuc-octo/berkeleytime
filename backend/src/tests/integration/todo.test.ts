import { createTestClient } from "apollo-server-testing";
import { ApolloServer, gql } from "apollo-server-express";
import { buildSchema } from "../../utils";
import mongoose from "mongoose";

import { resolvers } from "../../modules";
import { TodoMongooseModel } from "../../modules/todo/model";

import {
  connect,
  clearDatabase,
  closeDatabase,
  populateDatabase,
} from "../utils";

beforeAll(async () => connect());

// You can populate de DB before each test
beforeEach(async () => {
  await populateDatabase(TodoMongooseModel, [
    {
      content: "todo 1",
    },
    {
      content: "todo 2",
    },
  ]);
});

afterEach(async () => {
  await clearDatabase();
});

afterAll(async (done) => {
  await closeDatabase();
  done();
});

/**
 * Prompt test suite.
 */
describe("Todo", () => {
  it(`should create a todo`, async () => {
    // We build the schema
    const graphQLSchema = await buildSchema();

    // We create the Apollo Server
    const server = new ApolloServer({
      schema: graphQLSchema,
    }) as any;

    // use the test server to create a query function
    const { mutate } = createTestClient(server);

    // We define the query and the variables as you would do from your front-end
    const variables = {
      createTodoData: {
        content: `Test todo.`,
      },
    };

    const CREATE_TODO = gql`
      mutation createTodo($createTodoData: NewTodoInput!) {
        createTodo(createTodoData: $createTodoData) {
          content
        }
      }
    `;

    // run query against the server and snapshot the output
    const res = await mutate({
      mutation: CREATE_TODO,
      variables,
    });

    expect(res).toMatchSnapshot();
  });

  it(`should get a todo`, async () => {
    // We generate a todo ID
    const todoId = new mongoose.Types.ObjectId().toHexString().toString();

    // Add a todo with the generated ID in the database
    await populateDatabase(TodoMongooseModel, [
      {
        _id: todoId,
        content: "todo 1",
      },
    ]);

    // We build the schema
    const graphQLSchema = await buildSchema();

    // We create the Apollo Server
    const server = new ApolloServer({
      schema: graphQLSchema,
    }) as any;

    // use the test server to create a query function
    const { query } = createTestClient(server);

    const variables = {
      id: todoId,
    };

    const GET_TODO = gql`
      query getTodo($id: ObjectId!) {
        getTodo(id: $id) {
          content
        }
      }
    `;

    // run query against the server and snapshot the output
    const res = await query({
      query: GET_TODO,
      variables,
    });

    expect(res).toMatchSnapshot();
  });
});
