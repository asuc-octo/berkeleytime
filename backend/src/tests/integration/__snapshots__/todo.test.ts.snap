// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo should create a todo 1`] = `
Object {
  "data": Object {
    "createTodo": Object {
      "content": "Test todo.",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Todo should get a todo 1`] = `
Object {
  "data": Object {
    "getTodo": Object {
      "content": "todo 1",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
